### PR TITLE
New goto checker interfaces [blocks: 3579]

### PR DESCRIPTION
**Pull request by peterschrammel**
This will replace `safety_checkert` and `property_checkert`, and become the basis for breaking up the `bmct` monolith.
Based on https://github.com/diffblue/cbmc/pull/3557. Review only last commit.

I'm not happy with the naming of `goto_checkert`, which doesn't intuitively describe what it is doing - hoping for better suggestions from the reviewers.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: c54e65f66001e32fd9c58483b074f97726ffc976).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94498702
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 196891541cdbaa42197ca0bb6901edbcefc0c67e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94899319
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/goto-checker/goto_checker.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  /// \return whether the goto checker may HAVE_MORE results or
+  ///   whether it is DONE with the given properties.
+  /// After returning DONE, another call to operator() with the same
+  /// properties will return DONE and leave the properties' results unchanged.
+  /// If there is a property with status FAIL then its counterexample
+  /// can be retrieved by calling `build_error_trace` before any
+  /// subsequent call to operator().
+  /// `goto_checkert` derivatives shall be implemented in a way such
+  /// that repeated calls to operator() shall return the next FAILed
+  /// property until eventually not returning any FAILing properties.
```
Comment (romainbrenguier) [inline src/goto-checker/goto_checker.h]: This last sentence is not clear, since the operator does not return a property.
And this sounds like something we would want to be enforced by the interface.

File: src/goto-checker/goto_verifier.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include <util/options.h>
+#include <util/ui_message.h>
+
+#include "properties.h"
+
+/// An implementation of `goto_verifiert` checks all properties in
+/// a goto model. It typically uses, but doesn't have to use, a
+/// `goto_checkert` to iteratively determine the verification result
+/// of each property.
+class goto_verifiert
```
Comment (romainbrenguier) [inline src/goto-checker/goto_verifier.h:25]: I think checker and verifier can easily be confused and the difference in meaning is not clear. I also can't find a very good naming for this two classes. Maybe this class is the one that should be `goto_checker` and the other one could be named something like `incremental_goto_checker`? since it's only checking until the next counter-example is found.

File: src/goto-checker/properties.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  UNREACHABLE;
+}
+
+/// Returns the properties in the goto model
+propertiest initialize_properties(const abstract_goto_modelt &goto_model)
+{
+  propertiest properties;
+  forall_goto_functions(it, goto_model.get_goto_functions())
```
Comment (romainbrenguier) [inline src/goto-checker/properties.cpp]: :pick: an ranged-for would be preferred here, to make explicit that `it` is used just to access the data and not for other iterator operations. :rabbit2:

File: src/goto-checker/properties.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      const goto_programt &goto_program = it->second.body;
+
+      forall_goto_program_instructions(i_it, goto_program)
+      {
+        if(!i_it->is_assert())
+          continue;
+
+        const source_locationt &source_location = i_it->source_location;
+
+        irep_idt property_id = source_location.get_property_id();
```
Comment (romainbrenguier) [inline src/goto-checker/properties.cpp]: :pick: intermediary variable `source_location` is superfluous

Comment (romainbrenguier) [inline src/goto-checker/properties.cpp]: should be const

File: src/goto-checker/properties.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  /// An error occurred during goto checking
+  ERROR
+};
+
+std::string as_string(resultt result);
+
+struct property_infot
+{
+  /// A pointer to the corresponding goto instruction
+  goto_programt::const_targett location;
```
Comment (romainbrenguier) [inline src/goto-checker/properties.h]: The naming is a bit confusing as I would expect `location` to be a `source_locationt`. Could it be renamed to something like `instruction_pointer`?

**Review by romainbrenguier — APPROVED** [review]
Some naming suggestions, otherwise looks fine

File: src/goto-checker/goto_checker.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+/// \file
+/// Goto Checker Interface
+
+#ifndef CPROVER_GOTO_CHECKER_GOTO_CHECKER_H
+#define CPROVER_GOTO_CHECKER_GOTO_CHECKER_H
+
+#include <goto-programs/goto_trace.h>
+
+#include <util/options.h>
+#include <util/ui_message.h>
```
Comment (tautschnig) [inline src/goto-checker/goto_checker.h]: My preference is to use forward declarations (of `optionst` and `ui_message_handlert`) and just include `message.h` to reduce build dependencies.

File: src/goto-checker/goto_verifier.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+/// \file
+/// Goto Verifier Interface
+
+#ifndef CPROVER_GOTO_CHECKER_GOTO_VERIFIER_H
+#define CPROVER_GOTO_CHECKER_GOTO_VERIFIER_H
+
+#include <util/optional.h>
+#include <util/options.h>
+#include <util/ui_message.h>
```
Comment (tautschnig) [inline src/goto-checker/goto_verifier.h:17]: As above: I'd suggest forward declarations.

File: src/goto-checker/goto_checker.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+#include "properties.h"
+
+/// An implementation of `goto_checkert` provides functionality for
+/// checking a set of properties and returning counterexamples
+/// one by one to the caller.
+/// An implementation of `goto_checkert` is responsible for maintaining
+/// the state of the analysis that it performs (e.g. goto-symex, solver, etc).
+/// It is not responsible for maintaining outcomes (e.g. property results,
+/// counterexamples, etc). However, an implementation may restrict the
+/// sets of properties it is asked to check (e.g. only sequences of subsets).
```
Comment (tautschnig) [inline src/goto-checker/goto_checker.h]: "subsets" of what?

File: src/goto-checker/goto_checker.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+#include <util/ui_message.h>
+
+#include "properties.h"
+
+/// An implementation of `goto_checkert` provides functionality for
+/// checking a set of properties and returning counterexamples
+/// one by one to the caller.
+/// An implementation of `goto_checkert` is responsible for maintaining
+/// the state of the analysis that it performs (e.g. goto-symex, solver, etc).
+/// It is not responsible for maintaining outcomes (e.g. property results,
```
Comment (tautschnig) [inline src/goto-checker/goto_checker.h]: Does "It" refer to the implementation? I'd use "An implementation of `goto_checkert` is _not_ responsible for maintaining outcomes ..."

File: src/goto-checker/goto_verifier.h (hunk 2 above)
Comment (tautschnig) [reply src/goto-checker/goto_verifier.h:25]: +1 - I also believe, but am not really sure, that the incremental nature is the difference? If this is correct, then the naming proposed by @romainbrenguier seems good; if it isn't, then the documentation needs to be revisited.

File: src/goto-checker/goto_checker.h [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    HAVE_MORE,
+    /// The goto checker has returned all results for the given set
+    /// of properties.
+    DONE
+  };
+
+  /// Check whether the given properties with status UNKNOWN or newly
+  /// discovered properties hold
+  /// \param [out] properties: Updates those properties whose
+  ///   results have been determined. Newly discovered properties are added.
```
Comment (allredj) [inline src/goto-checker/goto_checker.h]: :pick: Argument descriptions should talk about the arguments themselves, not the function. Maybe rephrase to:
`Properties updated to whether their results have been determined`.

File: src/goto-checker/goto_checker.h [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  /// properties will return DONE and leave the properties' results unchanged.
+  /// If there is a property with status FAIL then its counterexample
+  /// can be retrieved by calling `build_error_trace` before any
+  /// subsequent call to operator().
+  /// `goto_checkert` derivatives shall be implemented in a way such
+  /// that repeated calls to operator() shall return the next FAILed
+  /// property until eventually not returning any FAILing properties.
+  virtual statust operator()(propertiest &properties) = 0;
+
+  /// This builds and returns the counterexample
```
Comment (allredj) [inline src/goto-checker/goto_checker.h]: :pick: Redundant `This`

File: src/goto-checker/properties.cpp [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  case property_resultt::FAIL:
+    return "FAIL";
+  case property_resultt::ERROR:
+    return "ERROR";
+  }
+
+  UNREACHABLE;
+}
+
+/// Returns the properties in the goto model
```
Comment (allredj) [inline src/goto-checker/properties.cpp]: :pick: `Returns` -> `Return`

File: src/goto-checker/goto_checker.h [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+/// counterexamples, etc). However, an implementation may restrict the
+/// sets of properties it is asked to check (e.g. only sequences of subsets).
+class goto_checkert
+{
+public:
+  enum class statust
+  {
+    /// The goto checker may be able to determine the results for more
+    /// properties if operator() is called again.
+    HAVE_MORE,
```
Comment (smowton) [inline src/goto-checker/goto_checker.h]: `PARTIAL_RESULT`?

File: src/goto-checker/goto_checker.h [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    /// The goto checker has returned all results for the given set
+    /// of properties.
+    DONE
+  };
+
+  /// Check whether the given properties with status UNKNOWN or newly
+  /// discovered properties hold
```
Comment (smowton) [inline src/goto-checker/goto_checker.h]: Newly discovered by the checker or by the caller?

File: src/goto-checker/goto_checker.h [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  /// \param [out] properties: Updates those properties whose
+  ///   results have been determined. Newly discovered properties are added.
+  /// \return whether the goto checker may HAVE_MORE results or
+  ///   whether it is DONE with the given properties.
+  /// After returning DONE, another call to operator() with the same
+  /// properties will return DONE and leave the properties' results unchanged.
+  /// If there is a property with status FAIL then its counterexample
+  /// can be retrieved by calling `build_error_trace` before any
```
Comment (smowton) [inline src/goto-checker/goto_checker.h]: This sounds like it binds an implementation into an odd stateful implementation, how about just passing a boolean to `operator()` to indicate whether you want traces built or not?

File: src/goto-checker/properties.h [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+#define CPROVER_GOTO_CHECKER_PROPERTIES_H
+
+#include <unordered_map>
+
+#include <goto-programs/goto_model.h>
+
+/// The result status of a property
+enum class property_resultt
+{
+  /// The property was not reached (also used for initialization)
```
Comment (smowton) [inline src/goto-checker/properties.h]: Not reached yet, I guess -- what's the difference between this and UNKNOWN?

**Comment by smowton** [comment]
Re: naming both the `checker` and the `verifier` sound like they might be checking internal consistency of a GOTO program, rather than checking whether its assertions can be violated. Like Romain though I don't have a better idea.

File: src/goto-checker/properties.h (hunk 16 above)
Comment (romainbrenguier) [reply src/goto-checker/properties.h]: It would rather be for unreachable assertions (for instance in methods that are loaded but not called from the entry point)

Comment (smowton) [reply src/goto-checker/properties.h]: That seems the same as `NOT_REACHABLE`. Perhaps they should be `NOT_REACHABLE` and `NOT_SATISFIABLE` (can get there, but can't concurrently satisfy the condition)?

File: src/goto-checker/properties.cpp [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+propertiest initialize_properties(const abstract_goto_modelt &goto_model)
+{
+  propertiest properties;
+  forall_goto_functions(it, goto_model.get_goto_functions())
+  {
+    if(!it->second.is_inlined() || it->first == goto_functionst::entry_point())
```
Comment (allredj) [inline src/goto-checker/properties.cpp]: :pick: Maybe document why we exclude all inlined functions except the entry point?

File: src/goto-checker/properties.cpp (hunk 12 above)
Comment (allredj) [reply src/goto-checker/properties.cpp]: :pick: And perhaps document that this function also initialises the properties (it's clear from the name, but it's confusing that the documentation doesn't match the function name).

**Review by allredj — APPROVED** [review]
It would be good to apply Romain's comments, especially about the ranged for.

File: src/goto-checker/properties.h (hunk 16 above)
Comment (peterschrammel) [reply src/goto-checker/properties.h]: I'll add examples to clarify these notions.

File: src/goto-checker/goto_checker.h (hunk 14 above)
Comment (peterschrammel) [reply src/goto-checker/goto_checker.h]: Both, unfortunately.

File: src/goto-checker/goto_checker.h (hunk 15 above)
Comment (peterschrammel) [reply src/goto-checker/goto_checker.h]: 1. That will not make it less stateful if you still have to call a function to retrieve the trace.
2. A bool `should_build_trace` together with `optional<goto_tracet>` has the problem of what to return when not buildlng a trace and the status is HAVE_MORE... an empty trace?
2. One could have two operators: one that returns `optional<goto_tracet>` and another one that returns a status. The caller can pick which one to use depending on whether they want traces or not.

File: src/goto-checker/properties.h (hunk 16 above)
Comment (peterschrammel) [reply src/goto-checker/properties.h]: Assume we run a path-symex that does at most 5 unwindings and stops
before exploring the else branch (*).

```c
int main(int argc, char **argv)
{
  if(argc < 0)
    assert(0); // NOT_REACHABLE
  assert(argc >= 0); // PASS
  if(argc < 5)
  {
    for(int i=0; i<10; i++)
    {
      assert(i == 2) // FAIL
      assert(i != 15); // UNKNOWN
    }
  }
  else // (*)
  {
    assert(0); // NOT_REACHED
  }
}
```

File: src/goto-checker/goto_checker.h (hunk 14 above)
Comment (smowton) [reply src/goto-checker/goto_checker.h]: Suggest adding that to the comment then

File: src/goto-checker/goto_checker.h (hunk 15 above)
Comment (smowton) [reply src/goto-checker/goto_checker.h]: I was envisaging something like `std::pair<statust, optionalt<tracet>>`, but option 3 sounds pretty good to me

File: src/goto-checker/properties.h (hunk 16 above)
Comment (smowton) [reply src/goto-checker/properties.h]: How about `NOT_REACHED` -> `NOT_CHECKED`? `NOT_REACHED` and `NOT_REACHABLE` are so close to synonyms it will surely confuse.

File: src/goto-checker/goto_checker.h (hunk 15 above)
Comment (peterschrammel) [reply src/goto-checker/goto_checker.h]: Hm... having to decide whether I want a trace or not before actually knowing which property failed is a limitation for certain algorithms (e.g. synthesising invariants while checking properties: I don't need a counterexample when I fail due to a non-inductive invariant candidate, but I'd like to have the counterexample when a property fails). 
So, I think the two operations should be separate. I'm not worried about the statefulness because it is quite natural.

File: src/goto-checker/goto_checker.h (hunk 6 above)
Comment (peterschrammel) [reply src/goto-checker/goto_checker.h]: Not for class members, though.

**Comment by tautschnig** [comment]
@peterschrammel Please send a ping when this one is ready for (re-)review.

File: src/goto-checker/properties.h (hunk 5 above)
Comment (peterschrammel) [reply src/goto-checker/properties.h]: Going with `pc` which is used in other classes such as `goto_tracet`.

File: src/goto-checker/goto_verifier.h (hunk 2 above)
Comment (peterschrammel) [reply src/goto-checker/goto_verifier.h:25]: Going with `incremental_goto_checkert` due to lack of more intuitive names.

File: src/goto-checker/properties.cpp (hunk 17 above)
Comment (peterschrammel) [reply src/goto-checker/properties.cpp]: This is a good question. It seems legacy to me and we should be able to remove this filter.
The first condition used to be necessary in times where `goto_partial_inline` was used.
The second condition is unclear to me (is the entry point marked inlined?). @kroening, @tautschnig, any insights?

File: src/goto-checker/goto_checker.h (hunk 8 above)
Comment (peterschrammel) [reply src/goto-checker/goto_checker.h]: Please check whether it's clearer now.

**Comment by peterschrammel** [comment]
> Please send a ping when this one is ready for (re-)review.

*ping*

File: src/goto-checker/properties.cpp (hunk 17 above)
Comment (tautschnig) [reply src/goto-checker/properties.cpp]: If the entry point were marked as inlined that would just seem wrong (and useless) to me. I'd suggest to remove this second part of the condition.

File: src/goto-checker/goto_verifier.h [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+#include "properties.h"
+
+/// An implementation of `goto_verifiert` checks all properties in
+/// a goto model. It typically uses, but doesn't have to use, a
+/// `goto_checkert` to iteratively determine the verification result
```
Comment (tautschnig) [inline src/goto-checker/goto_verifier.h]: Should this be `incremental_goto_checkert`?

File: src/goto-checker/goto_verifier.h [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+  /// Report results
+  virtual void report() = 0;
+
+  /// Returns the properties
+  const propertiest &get_properties()
+  {
+    return properties;
+  }
+
```
Comment (tautschnig) [inline src/goto-checker/goto_verifier.h:48]: Needs a `virtual ~goto_verifiert() = default;`

File: src/goto-checker/incremental_goto_checker.cpp [hunk 20]
```diff
@@ -0,0 +1,21 @@
+/*******************************************************************\
+
+Module: Goto Checker Interface
```
Comment (tautschnig) [inline src/goto-checker/incremental_goto_checker.cpp]: Nit-pick: add "Incremental"

File: src/goto-checker/incremental_goto_checker.h [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+{
+public:
+  enum class statust
+  {
+    /// The goto checker may be able to find another FAILed property
+    /// if operator() is called again.
+    FOUND_FAIL,
+    /// The goto checker has returned all results for the given set
+    /// of properties.
+    DONE
```
Comment (tautschnig) [inline src/goto-checker/incremental_goto_checker.h:47]: So what's the initial status to use then?

File: src/goto-checker/incremental_goto_checker.h [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+  /// can be retrieved by calling `build_error_trace` before any
+  /// subsequent call to operator().
+  /// `goto_checkert` derivatives shall be implemented in a way such
+  /// that repeated calls to operator() shall return when the next FAILed
+  /// property has been found until eventually it does not find any
+  /// failing properties any more.
+  virtual statust operator()(propertiest &properties) = 0;
+
+  /// Builds and returns the counterexample
+  virtual goto_tracet build_error_trace() const = 0;
```
Comment (tautschnig) [inline src/goto-checker/incremental_goto_checker.h:68]: Do we have move operators so that we can reasonably pass a `goto_tracet` by value?

File: src/goto-checker/properties.cpp [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+
+/// Return the properties in the goto model and initialize them to NOT_CHECKED
+propertiest initialize_properties(const abstract_goto_modelt &goto_model)
+{
+  propertiest properties;
+  const auto &goto_functions = goto_model.get_goto_functions();
+  for(const auto &function_pair : goto_functions.function_map)
+  {
+    if(!function_pair.second.is_inlined() ||
+      function_pair.first == goto_functionst::entry_point())
```
Comment (tautschnig) [inline src/goto-checker/properties.cpp]: See comment on the earlier version: the second condition should be removed.

File: src/goto-checker/properties.cpp [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+      // need pointer to goto instruction
+      forall_goto_program_instructions(i_it, goto_program)
+      {
+        if(!i_it->is_assert())
+          continue;
+
+        property_infot &property =
+          properties[i_it->source_location.get_property_id()];
+        property.result = property_resultt::NOT_CHECKED;
+        property.pc = i_it;
```
Comment (tautschnig) [inline src/goto-checker/properties.cpp]: A `property_infot` should have a constructor so that we can do
`properties.emplace(i_it->source_location.get_property_id(), {property_resultt::NOT_CHECKED, i_it});` here.

**Review by tautschnig — APPROVED** [review]

File: src/goto-checker/incremental_goto_checker.h (hunk 21 above)
Comment (peterschrammel) [reply src/goto-checker/incremental_goto_checker.h:47]: Why would you need an initial status?

File: src/goto-checker/incremental_goto_checker.h (hunk 22 above)
Comment (peterschrammel) [reply src/goto-checker/incremental_goto_checker.h:68]: Should be implicit

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f1f62b29081787043c41727b414d1b2ba0c2b0c3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97146777

File: src/goto-checker/incremental_goto_checker.h (hunk 21 above)
Comment (tautschnig) [reply src/goto-checker/incremental_goto_checker.h:47]: There always has to be some "status" - otherwise it's a "result" I'd say?

Comment (peterschrammel) [reply src/goto-checker/incremental_goto_checker.h:47]: Made naming of result vs status more consistent

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a1100e16be09af1e9278d8dd2521d3699cfefc63).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97160722
